iis-wf/iis-wf-import/src/main/java/eu/dnetlib/iis/wf/importer/infospace/ImportInformationSpaceJob.java

```diff
@@ -0,0 +1,404 @@
+
+        JavaRDD<eu.dnetlib.iis.importer.schemas.DocumentMetadata> softwareDocMeta = filterAndParseResultToAvro(
+                sourceSoftware, resultApprover, documentConverter);
+
+        JavaRDD<eu.dnetlib.iis.importer.schemas.DocumentMetadata> orpDocMeta = filterAndParseResultToAvro(
+                sourceOtherResearchProduct, resultApprover, documentConverter);
+
+        JavaRDD<eu.dnetlib.iis.importer.schemas.DocumentMetadata> datasetDocMeta = parseResultToAvro(
+                filteredDataset, documentConverter);
+
+        return publicationDocMeta.union(softwareDocMeta).union(orpDocMeta).union(datasetDocMeta);
+    }
+    
+    /**
+     * Parses given RDD with {@link OafEntity} entities into {@SpecificRecord} avro RDD.
+     */
+    private static <S extends OafEntity, T extends SpecificRecord>JavaRDD<T> filterAndParseToAvro(JavaRDD<S> source, 
+            ResultApprover resultApprover, OafEntityToAvroConverter<S, T> entityConverter) {
+        return parseToAvro(source.filter(x -> resultApprover.approve(x)), entityConverter);
+    }
+    
+    /**
+     * Converts given RDD with {@link OafEntity} entities into {@link SpecificRecord} avro RDD. Eliminates null records from output.
+     */
+    private static <S extends OafEntity, T extends SpecificRecord>JavaRDD<T> parseToAvro(JavaRDD<S> source, 
+            OafEntityToAvroConverter<S, T> entityConverter) {
+        return source.map(x -> entityConverter.convert(x)).filter(x -> x != null);
+    }
+    
+    /**
+     * Parses given RDD with {@link Result} entities into {@link SpecificRecord} avro RDD.
+     */
+    private static <T extends SpecificRecord>JavaRDD<T> filterAndParseResultToAvro(JavaRDD<? extends Result> source, 
+            ResultApprover resultApprover, OafEntityToAvroConverter<Result, T> entityConverter) {
+        return parseResultToAvro(source.filter(x -> resultApprover.approve(x)), entityConverter);
+    }
+
+    /**
+     * Converts given RDD with {@link Result} entities into {@link SpecificRecord} avro RDD. Eliminates null records from output.
+     */
+    private static <T extends SpecificRecord>JavaRDD<T> parseResultToAvro(JavaRDD<? extends Result> source, 
+            OafEntityToAvroConverter<Result, T> entityConverter) {
+        return source.map(x -> entityConverter.convert(x)).filter(x -> x != null);
+    }
+    
+    /**
+     * Converts given RDD with {@link Result} entities into {@link SpecificRecord} avro RDD. Eliminates null records from output.
+     */
+    private static <T extends SpecificRecord> JavaRDD<T> filterAndParseRelationToAvro(JavaRDD<Relation> source,
+            ResultApprover resultApprover, OafRelToAvroConverter<T> relationConverter, String relType,
+            String subRelType, String relClass) {
+        return source.filter(x -> acceptRelation(x, relType, subRelType, relClass))
+                .filter(x -> resultApprover.approve(x)).map(x -> relationConverter.convert(x));
```

Comment:
Hmm, this one is strange because in this particular case referencing the methods with `::` causes:
```
 java.lang.ClassCastException: eu.dnetlib.dhp.schema.oaf.Relation cannot be cast to eu.dnetlib.dhp.schema.oaf.OafEntity
```
I cannot find any reference to `OafEntity` neither in approver nor converter. It looks like the execution plan includes some invalid optimization or sth. Really weird one.

Similar replacements in other places in the code does not result in such behaviour.

Reply:
This is really strange, because there is no reference to `OafEntity` in the code :/